api/api/openapi.py

```diff
@@ -1,38 +1,68 @@
-
-    # Restrict segment rule recursion to two levels.
-    segment_rule_schema = deepcopy(model_json_schema["$defs"]["SegmentRuleModel"])
-    del segment_rule_schema["properties"]["rules"]
-    model_json_schema["$defs"]["SegmentRuleInnerModel"] = segment_rule_schema
-    model_json_schema["$defs"]["SegmentRuleModel"]["properties"]["rules"]["items"][
-        "$ref"
-    ] = "#/$defs/SegmentRuleInnerModel"
-
-    # Remove integrations.
-    for prop in SKIP_PROPERTIES:
-        del model_json_schema["properties"][prop]
-    for definition in SKIP_DEFINITIONS:
-        del model_json_schema["$defs"][definition]
-
-    return Schema(**jsonref.replace_refs(model_json_schema))
+from typing import Any
+
+from drf_yasg.inspectors import SwaggerAutoSchema
+from drf_yasg.openapi import SCHEMA_DEFINITIONS, Response, Schema
+from pydantic import BaseModel
+from pydantic.json_schema import GenerateJsonSchema, JsonSchemaValue
+from pydantic_core import core_schema
+
+
+class _GenerateJsonSchema(GenerateJsonSchema):
+    def nullable_schema(self, schema: core_schema.NullableSchema) -> JsonSchemaValue:
+        """Generates an OpenAPI 2.0-compatible JSON schema that matches a schema that allows null values.
+
+        Args:
+            schema: The core schema.
+
+        Returns:
+            The generated JSON schema.
+        """
+        anyof_schema_value = super().nullable_schema(schema)
+        elem = next(
+            any_of
+            for any_of in anyof_schema_value["anyOf"]
+            if any_of.get("type") != "null"
+        )
+        if type := elem.get("type"):
+            return {"type": type, "x-nullable": True}
+        # Assuming a reference here (which we can not annotate)
+        return elem
+
+
+class PydanticResponseCapableSwaggerAutoSchema(SwaggerAutoSchema):
+    def get_response_schemas(
+        self,
+        response_serializers: dict[str | int, Any],
+    ) -> dict[str, Response]:
+        result = {}
+
+        for status_code in list(response_serializers):
+            if isinstance(response_serializers[status_code], type) and issubclass(
+                model_cls := response_serializers[status_code], BaseModel
+            ):
+                model_json_schema = model_cls.model_json_schema(
+                    mode="serialization",
+                    schema_generator=_GenerateJsonSchema,
+                    ref_template=f"#/{SCHEMA_DEFINITIONS}/{{model}}",
+                )
+
+                definitions = self.components.with_scope(SCHEMA_DEFINITIONS)
+                for ref_name, schema_kwargs in model_json_schema.pop("$defs").items():
+                    definitions.setdefault(
+                        ref_name,
+                        maker=lambda: Schema(
+                            **schema_kwargs,
+                            # We can not annotate references with `x-nullable`,
+                            # So just assume all nested models as nullable for now.
```

Comment:
This is all quite abstract so I'm struggling to understand exactly where this limitation exists? I think this means that whenever we define schema in a swagger response schema definition as a pydantic model that has a nested pydantic model, it will always be shown as nullable regardless of the type definition?

Reply:
> exactly where this limitation exists?

Any properties sibling to `$ref` are ignored according to [OpenAPI spec](https://swagger.io/docs/specification/using-ref/) (see "$ref Syntax").

> it will always be shown as nullable regardless of the type definition

Correct. This, IMO, is fine for response schemas.